New hdf5 file format for saving PeaksAndMetrics objects

Garyfallidis opened the pull request:
Hi @matthieudumont and all this PR is resolving the issue with saving the peaks object. Making it an HDF5 file allows to be read by other languages etc.

Mr. Dumont you will need to update your workflow PRs to use this one and then we can move to merging them asap.

Do you agree with the file extension .pam or .pam5 (as in hdf5)?

Also have in mind that the helper peaks_to_niftis is not in this PR so you will have to use it from your other PR. Let me know how it goes.

arokem commented on doc/examples/introduction_to_basic_tracking.py:
```diff
@@ -66,6 +66,7 @@
 the ODF shows significant restricted diffusion by thresholding on
 the general fractional anisotropy (GFA).
 """
+1/0
```
This probably shouldn't be here anymore 😄

Garyfallidis replied:
Sure

arokem commented on doc/examples/viz_widgets.py:
```diff
@@ -121,9 +121,9 @@ def text_clicked(obj, event):
-# show_manager.initialize()
-# show_manager.render()
-# show_manager.start()
+show_manager.initialize()
```
These should probably be recommented

Garyfallidis replied:
Sure.

arokem replied:
Please don't reply to these comments, unless there is something to discuss. I get an email every time you make a comment like this one.

arokem commented on dipy/io/peaks.py:
```diff
@@ -0,0 +1,167 @@
+
+
+def load_peaks(fname, verbose=True):
+    """ Load PeaksAndMetrics PAM file (HDF5)
```
Docstring?

arokem commented on dipy/io/peaks.py:
```diff
@@ -0,0 +1,167 @@
+
+
+def save_peaks(fname, pam):
+    """ Save NPZ file with all important attributes of object PeaksAndMetrics
```
Docstring

arokem commented:
Why not use just use the `hdf5` extension?

arokem commented on dipy/io/tests/test_io_peaks.py:
```diff
@@ -0,0 +1,63 @@
+
+        save_peaks(fname, pam)
+        pam2 = load_peaks(fname, verbose=True)
+        npt.assert_array_equal(pam.peak_dirs, pam2.peak_dirs)
```
Why not test that all the attributes survived the round-trip?

arokem replied:
This comment is still relevant.

arokem replied:
I think this comment is still relevant?

Garyfallidis commented:
Hi @arokem. The reason why I suggest not using the extension hdf5 is that hdf5 is too generic and then .pam_hdf5 is is too long. Also, we are using hdf5 to store streamlines too in the .dpy format. This is used by quite some people who were really unpleased with trk format (for a good reason). However, .pam is already used by netpbm as an image format. So, I suggest to use .pam5 instead. I think with the clue of the 5 and some info in the docstring it will be difficult for people to miss that this is actually an hdf5 file.
Agree? Also if we use the convention ``_pam.hdf5`` we violate the extension rules. I mean the file format extension definition should be after the last dot.

arokem commented:
As you say, hdf5 is generic. That is also an advantage. For example, imagine that someone created an application that is a tool for visualizing the content of an hdf5 file (just an example, it can be other things). This kind of application might not work on these files for prosaic reasons, because the application wouldn't 'see' these files as hdf5, even though that's what they are.

If you insist on creating a new file format, I suggest versioning it and checking for the version. It's possible that you will want to change this in the future and you might run into cryptic incompatibilities unless you check for the version of the file. 

But I really wouldn't hurry to proliferate a new standard -- seems like something that should be done only when there is  clear necessity for it, which I don't see here.

Garyfallidis commented:
There is a necessity for the following reasons:
 
a) If we do use a different extension name then it is becomes easier for our visualization software to read the data. Is it clear why? The viz software will check the file extension and then directly call load_peaks (from dipy.viz). Otherwise if the file is .hdf it will have first to open the file and then see if it has the expected attributes and then see if it has loaded other datasets of .hdf files that are of different type. This makes the visualization duplicate effort with IO which I think is a bad idea.

b) We really need a file to store results between the reconstruction and tracking this has been a bottleneck for some time. Recently we started pickling the pam objects which was not possible before. But hdf5 allows us to use memory maps too and it should be faster than pickling too.

c) External software can build a similar functionality as we could potentially do in dipy.viz
There is also a function that we have implemented with Mr. Dumont that allows to extract some important arrays from PeaksAndMetrics to niftis to help with visualization tools that only read nifti files. This functionality is still on Mr. Dumont's PR.

Nonetheless I think we can both be happy if I use the name .pam_hdf5 rather than .pam5?
Adding a version is good idea too. I am happy to do that.
Let me know what you think.

arokem commented:
On Mon, May 1, 2017 at 1:37 PM, Eleftherios Garyfallidis <
notifications@github.com> wrote:

> There is a necessity for the following reasons:
>
> a) If we do use a different extension name then it is becomes easier for
> our visualization software to read the data. Is it clear why? The viz
> software will check the file extension and then directly call load_peaks
> (from dipy.viz). Otherwise if the file is .hdf it will have first to open
> the file and then see if it has the expected attributes and then see if it
> has loaded other datasets of .hdf files that are of different type. This
> makes the visualization duplicate effort with IO which I think is a bad
> idea.
>
OK - this seems like an argument in favor of a new file format. Of course,
if users always feed these hdf5 files in, that's probably not a problem.
But I agree that having visualization software read files in and check them
wouldn't be great. Is there some way to read an hdf5 file's header for some
pertinent meta-data?

> b) We really need a file to store results between the reconstruction and
> tracking this has been a bottleneck for some time. Recently we started
> pickling the pam objects which was not possible before. But hdf5 allows us
> to use memory maps too and it should be faster than pickling too.
>
This seems irrelevant to the discussion. When you pickle the pam objects,
do you store them in a different file-format (I'm guessing not)?

> c) External software can build a similar functionality as we could
> potentially do in dipy.viz
>
It's more likely that external software would build functionality for a
well-established file format, than for our newly designed file format.

> There is also a function that we have implemented with Mr. Dumont that
> allows to extract some important arrays from PeaksAndMetrics to niftis to
> help with visualization tools that only read nifti files. This
> functionality is still on Mr. Dumont's PR.
>
This actually seems like an argument for using existing file formats. You
wouldn't design a new `nipam` file format for these, right? ;-)

> Nonetheless I think we can both be happy if I use the name .pam_hdf5
> rather than .pam5?
>
How about a double extension: `.pam.hdf5`? That way, external software will
recognize this as an hdf5 file, and our software can look for the full
`.pam.hdf5` to make sure that this is the right file.

> Adding a version is good idea too. I am happy to do that.
> Let me know what you think.
>
:+1:


> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/nipy/dipy/pull/1235#issuecomment-298425818>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAHPNhudwOouekyS11tUxDaOQM_1hWYVks5r1kJ9gaJpZM4NMd3s>
> .
>

Garyfallidis commented:
Added version and improved testing. Waiting for travis. On the name I do prefer .pam5 than .pam.hdf5

arokem commented:
Why do you prefer `.pam5` over `.pam.hdf5`? The latter seems to cover all the use-cases described above, while the former will fail at least for the case of software written for hdf5 files by people unaware of the `pam5` format.

Garyfallidis commented:
We do use the two dots for example when we say that a file has two encodings for example .tar.gz and .nii.gz but those are actually two different processing levels in both cases. In our case there is only one processing level. This is one reason why I do not like the .pam.hdf5. Pam is an already available fileformat (for images). The second reason is that people will need to be checking for two last extensions rather than one. Which is not slick. 
Now, .pam_hdf5 is too long. So a shortened version of that is .pam5 which is short and unique.
Alternatively we can use .peaks but I would keep that for when the fit is separated from the peaks_and_models. But I am fine with .peaks too if you prefer that.

Garyfallidis commented:
About something else. The coverage I am getting in my machine for dipy/io/peaks.py is 96% but one of the bots is reporting 9%. Similarly it reports very low coverage for dipy/io/dpy.py 24% when in my machine I am getting 90%.

arokem commented:
That's because the coverage machine doesn't have pytables installed:
https://travis-ci.org/nipy/dipy/jobs/229548518#L2313

On Sat, May 6, 2017 at 6:49 PM, Eleftherios Garyfallidis <
notifications@github.com> wrote:

> About something else. The coverage I am getting in my machine for
> dipy/io/peaks.py is 96% but one of the bots is reporting 9%. Similarly it
> reports very low coverage for dipy/io/dpy.py 24% when in my machine I am
> getting 90%.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/nipy/dipy/pull/1235#issuecomment-299676817>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAHPNqmsrJqksh7cz3vh1J8a9CIOpmQ_ks5r3SMxgaJpZM4NMd3s>
> .
>

Garyfallidis commented:
Yes, it skips the test but still gives coverage for the file. Okay let me add pytables in.

arokem commented:
Yes. It gives coverage for everything that it can import from dipy.

Garyfallidis commented:
Travis and coveralls look happy. :)

Garyfallidis commented:
Now trying to reach maximum possible coverage.

arokem commented:
Regarding file format: to be clear, all the options that are new file
format extensions that we invent for this purpose seem almost equally bad
to me. I agree that `pam5` is the least offensive, but it's still a new
file format, and I just don't see the justification to create an entirely
new file format for what is essentially an hdf5 file.

The dual extension solution lets us eat our cake and have it too, but you
don't seem to like that. I am not sure why, though. Because in other cases
the two extensions represent different parts of the processing? Arguably
that's true here as well, one part reads the HDF5 file, and the other
parses it for PAM attributes.




On Sun, May 7, 2017 at 10:49 AM, Eleftherios Garyfallidis <
notifications@github.com> wrote:

> Now trying to reach maximum possible coverage.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/nipy/dipy/pull/1235#issuecomment-299722471>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAHPNtoGU5yUFeJ68EZztFTIAX9YMMTgks5r3gQxgaJpZM4NMd3s>
> .
>

Garyfallidis commented:
What I do not like is that one needs to parse for two extensions rather than one. Also the filename size is increasing. But more importantly .pam is already reserved for another popular format. So .pam.hdf is misleading. It's like we put a picture file (.pam) to an .hdf5 file. I hope this clarifies more.

arokem commented:
Yes - the latter is a good point. What picture file format uses `pam`? I
found this file format that uses pam: http://filext.com/file-extension/PAM.
Seems rather obscure.

On Sun, May 7, 2017 at 11:14 AM, Eleftherios Garyfallidis <
notifications@github.com> wrote:

> What I do not like is that one needs to parse for two extensions rather
> than one. Also the filename size is increasing. But more importantly .pam
> is already reserved for another popular format. So .pam.hdf is misleading.
> It's like we put a picture file (.pam) to an .hdf5 file. I hope this
> clarifies more.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/nipy/dipy/pull/1235#issuecomment-299724090>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAHPNgOr0kf7hLfRw9u0fjAyBmbFJW2Vks5r3goPgaJpZM4NMd3s>
> .
>

arokem commented:
On the other hand, `dpy` also seems to be used by others:
http://www.liutilities.com/products/winbackup/filextlibrary/files/DPY/

On Sun, May 7, 2017 at 11:27 AM, Ariel Rokem <arokem@gmail.com> wrote:

> Yes - the latter is a good point. What picture file format uses `pam`? I
> found this file format that uses pam: http://filext.com/file-extension/PAM.
> Seems rather obscure.
>
> On Sun, May 7, 2017 at 11:14 AM, Eleftherios Garyfallidis <
> notifications@github.com> wrote:
>
>> What I do not like is that one needs to parse for two extensions rather
>> than one. Also the filename size is increasing. But more importantly .pam
>> is already reserved for another popular format. So .pam.hdf is misleading.
>> It's like we put a picture file (.pam) to an .hdf5 file. I hope this
>> clarifies more.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/nipy/dipy/pull/1235#issuecomment-299724090>, or mute
>> the thread
>> <https://github.com/notifications/unsubscribe-auth/AAHPNgOr0kf7hLfRw9u0fjAyBmbFJW2Vks5r3goPgaJpZM4NMd3s>
>> .
>>
>
>

Garyfallidis commented:
It's part of the pbm suite of file formats http://netpbm.sourceforge.net/doc/pam.html

I made .dpy back in 2009. There was nothing with that extension at the time but also we were at the very early stages of development. I am happy to change that extension in the near future. I think we should but this time promote the tracking format. We didn't promote it enough so others used the same extension name.

Garyfallidis commented:
If we can merge this then I can go and update the workflow PRs and move ahead with release. 
All is left now is the workflows, new website and local pca right? I don't remember something else that was crucial for this release. But maybe I am wrong.

arokem commented:
Also these: 

https://github.com/nipy/dipy/issues/1178
https://github.com/nipy/dipy/pull/1164

arokem commented on .travis.yml:
```diff
@@ -53,6 +53,7 @@ matrix:
         - LIBGL_ALWAYS_INDIRECT=y
         - VENV_ARGS="--system-site-packages --python=/usr/bin/python2.7"
         - TEST_WITH_XVFB=true
+        - DEPENDS="$DEPENDS tables"
```
Add this on line 20, with the rest of the dependencies, instead of here.

Garyfallidis replied:
sure

arokem commented on dipy/io/peaks.py:
```diff
@@ -0,0 +1,237 @@
+
+    if verbose:
+        print('PAM5 version')
+        print(version)
```
I think that we should test up-front for the version, and raise an error if a file with the wrong version(s) is provided.

Garyfallidis replied:
Agree!

arokem commented on dipy/io/peaks.py:
```diff
@@ -0,0 +1,237 @@
+    Parameters
+    ----------
+    fname : string
+        Filenam of PAM5 file
```
typo: "Filenam" => "Filename". You might also want to put periods at the end of the sentences that describe parameters (here and elsewhere in these docstrings).

Garyfallidis replied:
Sure.

arokem commented on dipy/io/peaks.py:
```diff
@@ -0,0 +1,237 @@
+    if not (hasattr(pam, 'peak_dirs') and hasattr(pam, 'peak_values') and
+            hasattr(pam, 'peak_indices')):
+
+        raise ValueError('Cannot save object without peak_dirs/values/indices')
```
This error string might be confusing. How about: "Cannot save object without peak_dirs, peak_values and peak_indices" (to make it clear that all three are required).

Garyfallidis replied:
sure

arokem commented on doc/examples/viz_widgets.py:
```diff
@@ -118,11 +118,11 @@ def text_clicked(obj, event):
-# show_manager.initialize()
-# show_manager.render()
+show_manager.initialize()
+show_manager.render()
```
Do these still needs to be recommented, or is it just the line with `start` that will block?

Garyfallidis replied:
No, it does need to be commented because it can stop the execution of some files.

arokem commented:
BTW - where does the new website stand? We still had this issue related to that: https://github.com/nipy/dipy/issues/1197, and it doesn't seem like there's much progress on that...

Garyfallidis commented:
All good comments. Will get back to you later tonight.

arokem commented:
Did you know that I get emails when you make commits? That way you don't
have to say "sure" and "ok" every time you fix something. Github simply
sends me an email when you put in the commit. You can also put in a comment
in the thread when you are done addressing all the comments, asking me to
take another look. This is also helpful for me, because it minimizes the
amount of emails I get from Github. Thanks!

On Tue, May 9, 2017 at 5:59 PM, Eleftherios Garyfallidis <
notifications@github.com> wrote:

> *@Garyfallidis* commented on this pull request.
> ------------------------------
>
> In dipy/io/peaks.py
> <https://github.com/nipy/dipy/pull/1235#discussion_r115635887>:
>
> > +        Object holding peak_dirs, shm_coeffs and other attributes
> +    affine : array
> +        The 4x4 matrix transforming the date from native to world coordinates.
> +        PeaksAndMetrics should have that attribute but if not it can be
> +        provided here. Default None.
> +    verbose : bool
> +        Print summary information about the saved file.
> +    """
> +
> +    if os.path.splitext(fname)[1] != '.pam5':
> +        raise IOError('This function saves only PAM5 (HDF5) files')
> +
> +    if not (hasattr(pam, 'peak_dirs') and hasattr(pam, 'peak_values') and
> +            hasattr(pam, 'peak_indices')):
> +
> +        raise ValueError('Cannot save object without peak_dirs/values/indices')
>
> sure
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/nipy/dipy/pull/1235#discussion_r115635887>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAHPNuBdmDGQwnjM4fPWyo2PNI9X0_JUks5r4QwHgaJpZM4NMd3s>
> .
>

Garyfallidis commented:
The website is moving ahead. I made an account in a server her for Bishakh to run the daily doc builds, this a temporary solution and working with future systems https://portal.futuresystems.org to have a permanent solution for building the docs.

Garyfallidis commented:
About the reply strategy. I am using inbox (by gmail) and all dipy commit mails are stacked together automatically. I like replying directly to a comment so that you can see that I have addressed it. That is for making the reviewers' life easier.  It's some kind of reminder in a way for keeping track of the discussion. I mean github put a reply button there for a reason. Right? It's also giving some credit to the reviewer. Now, I could start putting all the short answers to the overall comment but that would make it difficult to connect to which lines these are referring.

Garyfallidis commented:
Inbox (by gmail) has also some smart notification system that sends you notifications only for crucial e-mails and ignores github e-mails. Anyway, I can try to reduce my short answers and you may want to relax your notification system.

arokem commented:
OK -- if it helps you track what you did, that's fine. I can always have
more email (it seems...)

I just thought that you thought it helps me see what's done, so I thought I
would mention that.

Are you done with this PR? Ready for another review?


On Tue, May 9, 2017 at 6:50 PM, Eleftherios Garyfallidis <
notifications@github.com> wrote:

> Inbox (by gmail) has also some smart notification system that sends you
> notifications only for crucial e-mails and ignores github e-mails. Anyway,
> I can try to reduce my short answers and you may want to relax your
> notification system.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/nipy/dipy/pull/1235#issuecomment-300350964>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAHPNrumY1mJvul4YEEraybAtytOH50cks5r4Rf2gaJpZM4NMd3s>
> .
>

Garyfallidis commented:
Okay good. Now travis is complaining on dki which is irrelevant to this PR. Interesting ...

Garyfallidis commented:
Sure, you can continue with the review. Please do check that I correctly change the travis configuration file. Now that tables is on the top shouldn't we remove it from the DEPENDS sections?

arokem commented:
I think that bug should be fixed in #1231.

arokem commented:
Regarding pytables: the question is whether we want it installed on all the machines? I think that since it is an optional install, the answer is "no". But it seems that you want it installed at least on the machine that checks coverage. Maybe remove it from the global depends and add both scikit learn and tables to the DEPENDS variable on the one machine that checks coverage? I would make two coverage machines: one with Python 3.5, and the other with Python 2.7. The Python 3.5 machine would have pytables and scikit learn installed. Meanwhile, the Python 2.7 machine would have both of these, as well as VTK. How's that?

Garyfallidis commented:
Hi, I think after this release we should actually start having tables, sklearn, cvxpy and vtk as not optional.
For now, I updated .travis.yml and rebased. Waiting for travis to see what we are getting. Added coverage for 3.5 and 2.7 as suggested.

Garyfallidis commented:
Hi Ariel, can you look into the change I made in travis for some reason the coverage was decreased after rebasing. This seems to be unrelated to the focus of this PR. And we may want to put a not to increase coverage but move on with the merging this and updating the workflow PRs. Let me know what you think.

arokem commented:
Yeah - I don't understand what coveralls wants. Maybe this has to do with
the addition of the Python 3 machines?

I had one tiny comment left: test all the attributes in the object for
round-tripping.  Other than that, this looks ready to go from my point of
view, so if anyone else wants to take a look, now is the time.

On Wed, May 10, 2017 at 6:20 PM, Eleftherios Garyfallidis <
notifications@github.com> wrote:

> Hi Ariel, can you look into the change I made in travis for some reason
> the coverage was decreased after rebasing. This seems to be unrelated to
> the focus of this PR. And we may want to put a not to increase coverage but
> move on with the merging this and updating the workflow PRs. Let me know
> what you think.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/nipy/dipy/pull/1235#issuecomment-300655854>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAHPNlLJ7MRUhusHubvds6z__fBjylLhks5r4mJPgaJpZM4NMd3s>
> .
>

Garyfallidis commented:
I am happy to test the attributes more. But what do you mean with round-tripping? Do you mean rounding errors?

arokem commented:
Sorry - with "round trip", I mean from memory to file and back. That is,
create the object, save the file, read the file into a new object and then
iterate over all of its attributes, checking that they are all identical to
the ones that came from the file.

Something like this:
https://github.com/nipy/dipy/blob/master/dipy/direction/tests/test_peaks.py#L507-L512

I think you are already checking a few of the attributes, and I would like
to check all of them.

On Thu, May 11, 2017 at 3:41 PM, Eleftherios Garyfallidis <
notifications@github.com> wrote:

> I am happy to test the attributes more. But what do you mean with
> round-tripping? Do you mean rounding errors?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/nipy/dipy/pull/1235#issuecomment-300936431>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAHPNjjvqX3Anj2BWi-LPXP86Nrvewy_ks5r446xgaJpZM4NMd3s>
> .
>

Garyfallidis commented:
Thanks for this tip. Good idea! Done and waiting for travis.
